### PR TITLE
Improve `ArrowWriter` memory usage: Buffer Pages in ArrowWriter instead of RecordBatch (#3871)

### LINKED ISSUE #3871: Avoid Buffering Arrow Data for Entire Row Group in parquet::ArrowWriter
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

Currently ArrowWriter buffers up `RecordBatch` until it has enough rows to populate an entire row group, and then proceeds to write each column in turn to the output buffer.

**Describe the solution you'd like**

The encoded parquet data is often orders of magnitude smaller than the corresponding arrow data. The read path goes to great lengths to allow incremental reading of data within a row group. It may therefore be desirable to instead encode arrow data eagerly, writing each ColumnChunk to its own temporary buffer, and then stitching these back together.

This would allow writing larger row groups, whilst potentially consuming less memory in the arrow writer.

This would likely involve extending or possibly replacing `SerializedRowGroupWriter` to allow writing to the same column multiple times

**Describe alternatives you've considered**

We could not do this, parquet is inherently a read-optimised format and write performance may therefore be less of a priority for many workloads.

**Additional context**

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #3871

# Rationale for this change
 

Buffering pages should be more memory efficient, and also allows for finer-grained control over row group sizing.

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: parquet/src/arrow/arrow_writer/byte_array.rs [hunk 1]
```diff
@@ -437,11 +390,11 @@ struct ByteArrayEncoder {
 
 impl ColumnValueEncoder for ByteArrayEncoder {
     type T = ByteArray;
-    type Values = ArrayRef;
+    type Values = dyn Array;
```
Comment (tustvold) [inline parquet/src/arrow/arrow_writer/byte_array.rs:393]: This is just a drive-by cleanup

File: parquet/src/arrow/arrow_writer/mod.rs [hunk 2]
```diff
@@ -152,43 +151,69 @@ impl<W: Write> ArrowWriter<W> {
...
 
-        for (buffer, column) in self.buffer.iter_mut().zip(batch.columns()) {
-            buffer.push_back(column.clone())
+        // If would exceed max_row_group_size, split batch
+        if self.buffered_rows + batch.num_rows() > self.max_row_group_size {
+            let to_write = self.max_row_group_size - self.buffered_rows;
+            let a = batch.slice(0, to_write);
+            let b = batch.slice(to_write, batch.num_rows() - to_write);
+            self.write(&a)?;
+            return self.write(&b);
```
Comment (tustvold) [inline parquet/src/arrow/arrow_writer/mod.rs:182]: I was being a bit lazy here, but this recursion will only be a problem if the the size of the RecordBatch is orders of magnitude larger than the row group size, which would be a pretty strange situation

File: parquet/src/arrow/arrow_writer/mod.rs [hunk 3]
```diff
@@ -152,43 +151,69 @@ impl<W: Write> ArrowWriter<W> {
...
-        self.flush_rows(self.buffered_rows)
+        let in_progress = match self.in_progress.take() {
+            Some(in_progress) => in_progress,
+            None => return Ok(()),
+        };
+
+        self.buffered_rows = 0;
+        let mut row_group_writer = self.writer.next_row_group()?;
+        for (chunk, close) in in_progress.close()? {
+            row_group_writer.append_column(&chunk, close)?;
```
Comment (tustvold) [inline parquet/src/arrow/arrow_writer/mod.rs:213]: This is the new API from #4269

File: parquet/src/arrow/arrow_writer/mod.rs [hunk 4]
```diff
@@ -284,156 +247,271 @@ impl<W: Write> RecordBatchWriter for ArrowWriter<W> {
...
+    fn len(&self) -> u64 {
+        self.length as _
+    }
+}
+
+impl ChunkReader for ArrowColumnChunk {
+    type T = ChainReader;
+
+    fn get_read(&self, start: u64) -> Result<Self::T> {
+        assert_eq!(start, 0);
```
Comment (tustvold) [inline parquet/src/arrow/arrow_writer/mod.rs:267]: We are somewhat relying on an implementation detail of append_column, but given this is the same crate, I think it is fine. It will break noisily should we change it

File: parquet/src/column/page.rs [hunk 5]
```diff
@@ -162,6 +162,75 @@ impl CompressedPage {
     pub fn data(&self) -> &[u8] {
         self.compressed_page.buffer().data()
     }
+
+    /// Returns the thrift page header
+    pub(crate) fn to_thrift_header(&self) -> PageHeader {
```
Comment (tustvold) [inline parquet/src/column/page.rs:167]: This is just moved from SerializedPageWriter

File: parquet/src/column/writer/encoder.rs [hunk 6]
```diff
@@ -36,7 +36,7 @@ pub trait ColumnValues {
 }
 
 #[cfg(feature = "arrow")]
-impl<T: arrow_array::Array> ColumnValues for T {
+impl ColumnValues for dyn arrow_array::Array {
```
Comment (tustvold) [inline parquet/src/column/writer/encoder.rs:39]: This trait is not public, so this isn't a breaking change

File: parquet/src/arrow/arrow_writer/mod.rs [hunk 7]
```diff
@@ -284,156 +247,271 @@ impl<W: Write> RecordBatchWriter for ArrowWriter<W> {
...
-        | ArrowDataType::Decimal256(_, _)
-        | ArrowDataType::FixedSizeBinary(_) => {
-            let mut col_writer = row_group_writer.next_column()?.unwrap();
-            for (array, levels) in arrays.iter().zip(levels.iter_mut()) {
-                write_leaf(col_writer.untyped(), array, levels.pop().expect("Levels exhausted"))?;
+/// A list of [`Bytes`] comprising a single column chunk
+#[derive(Default)]
+struct ArrowColumnChunk {
+    length: usize,
+    data: Vec<Bytes>,
```
Comment (tustvold) [inline parquet/src/arrow/arrow_writer/mod.rs:254]: By buffering separate chunks, we should be performing no more memcpy of data than we were before :tada:

File: parquet/src/column/writer/mod.rs [hunk 8]
```diff
@@ -1512,7 +1528,7 @@ mod tests {
             metadata.encodings(),
             &vec![Encoding::PLAIN, Encoding::RLE, Encoding::RLE_DICTIONARY]
         );
-        assert_eq!(metadata.num_values(), 8); // dictionary + value indexes
+        assert_eq!(metadata.num_values(), 4);
```
Comment (tustvold) [inline parquet/src/column/writer/mod.rs:1531]: I'm fairly certain this was just wrong before

**Comment by tustvold** [comment]
Benchmarks show no discernible impact on performance

File: parquet/src/arrow/arrow_writer/mod.rs [hunk 9]
```diff
@@ -152,43 +151,69 @@ impl<W: Write> ArrowWriter<W> {
         self.writer.flushed_row_groups()
     }
 
-    /// Enqueues the provided `RecordBatch` to be written
+    /// Returns the length in bytes of the current in progress row group
+    pub fn in_progress_size(&self) -> usize {
```
[the remark this answers is not in the document]
Comment (alamb) [reply parquet/src/arrow/arrow_writer/mod.rs:155]: I recommend we call this `in_progress_memory_size()` to be more consistent with `RecordBatch::get_array_memory_size()`

Also, could you please add an accessor for `buffered_rows`? perhaps

```rust
    pub fn in_progress_rows(&self) -> usize {
```

(I can add this as a follow on PR too if you prefer)

File: parquet/src/arrow/arrow_writer/mod.rs [hunk 10]
```diff
@@ -152,43 +151,69 @@ impl<W: Write> ArrowWriter<W> {
...
+    }
+
+    /// Encodes the provided [`RecordBatch`]
     ///
-    /// If following this there are more than `max_row_group_size` rows buffered,
-    /// this will flush out one or more row groups with `max_row_group_size` rows,
-    /// and drop any fully written `RecordBatch`
+    /// If this would cause the current row group to exceed [`WriterProperties::max_row_group_size`]
+    /// rows, the contents of `batch` will be distributed across multiple row groups such that all
+    /// but the final row group in the file contain [`WriterProperties::max_row_group_size`] rows
```
Comment (alamb) [inline parquet/src/arrow/arrow_writer/mod.rs:170]:
```suggestion
    /// If this would cause the current row group to exceed [`WriterProperties::max_row_group_size`]
    /// rows, the contents of `batch` will be written to one or more row groups such that all
    /// but the final row group in the file contain [`WriterProperties::max_row_group_size`] rows
```

(minor suggestion that I think is clearer -- feel free to ignore)

File: parquet/src/arrow/arrow_writer/mod.rs (hunk 2 above)
Comment (alamb) [reply parquet/src/arrow/arrow_writer/mod.rs:182]: I agree -- perhaps we can fix it if this turns out to be a problem in practice.

There is also a workaround for anyone who hits this, which is to break up their input RecordBatch (using `slice()`) into smaller parts

File: parquet/src/arrow/arrow_writer/mod.rs [hunk 11]
```diff
@@ -284,156 +247,271 @@ impl<W: Write> RecordBatchWriter for ArrowWriter<W> {
...
+}
+
+/// Encodes [`RecordBatch`] to a parquet row group
+struct ArrowRowGroupWriter {
+    writers: Vec<(SharedColumnChunk, ArrowColumnWriter)>,
+    schema: SchemaRef,
+}
+
+impl ArrowRowGroupWriter {
+    fn new(
```
Comment (alamb) [inline parquet/src/arrow/arrow_writer/mod.rs:362]: Given this returns `Result` I think `try_new(...)` would be a more idomatic name

File: parquet/src/arrow/arrow_writer/mod.rs [hunk 12]
```diff
@@ -152,43 +151,69 @@ impl<W: Write> ArrowWriter<W> {
...
+            let b = batch.slice(to_write, batch.num_rows() - to_write);
+            self.write(&a)?;
+            return self.write(&b);
         }
 
         self.buffered_rows += batch.num_rows();
-        self.flush_completed()?;
+        let in_progress = match &mut self.in_progress {
+            Some(in_progress) => in_progress,
+            x => x.insert(ArrowRowGroupWriter::new(
```
Comment (alamb) [inline parquet/src/arrow/arrow_writer/mod.rs:188]: TIL `insert` 👍  very nice

File: parquet/src/arrow/arrow_writer/mod.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (alamb) [reply parquet/src/arrow/arrow_writer/mod.rs:155]: Alternately, perhaps the code would be simpler if you removed `self.buffered_rows` and calculated it based on `in_progress`?

Perhaps something like 🤔 

```rust
    pub fn buffered_rows(&self) -> usize {
      self.in_progress.map(|p| p.len()).unwrap_or(0)
    }
```

File: parquet/src/arrow/arrow_writer/mod.rs (hunk 4 above)
Comment (alamb) [reply parquet/src/arrow/arrow_writer/mod.rs:267]: Maybe you can add a `//` comment here explaining what this is relying on, in case anyone in the future hits this assert

File: parquet/src/arrow/arrow_writer/mod.rs [hunk 13]
```diff
@@ -284,156 +247,271 @@ impl<W: Write> RecordBatchWriter for ArrowWriter<W> {
...
+        let b = buffer.split_to(len);
+        out[..len].copy_from_slice(&b);
+        Ok(len)
+    }
+}
+
+/// A shared [`ArrowColumnChunk`]
+///
+/// This allows it to be owned by [`ArrowPageWriter`] whilst allowing access via
+/// [`ArrowRowGroupWriter`] on flush, without requiring self-referential borrows
```
Comment (alamb) [inline parquet/src/arrow/arrow_writer/mod.rs:302]: I looked at this a bit -- is the issue that you can't get the `ColumnChunk` back from the `GenericColumnWriter`?

I wonder if we could add something like `GenericColumnWriter::close_into_inner()` that returned the inner writer and the ColumnCloseResult 🤔  

But that would still require downcasting back so 🤷  not sure it is any better

File: parquet/src/arrow/arrow_writer/mod.rs [hunk 14]
```diff
@@ -471,51 +549,45 @@ fn write_leaf(
...
         ColumnWriter::Int96ColumnWriter(ref mut _typed) => {
             unreachable!("Currently unreachable because data type not supported")
         }
         ColumnWriter::FloatColumnWriter(ref mut typed) => {
-            let array = column
-                .as_any()
-                .downcast_ref::<arrow_array::Float32Array>()
-                .expect("Unable to get Float32 array");
-            write_primitive(typed, array.values(), levels)?
+            let array = column.as_primitive::<Float32Type>();
```
Comment (alamb) [inline parquet/src/arrow/arrow_writer/mod.rs:585]: 🧹  nice and tidy

File: parquet/src/column/writer/mod.rs (hunk 8 above)
Comment (alamb) [reply parquet/src/column/writer/mod.rs:1531]: I am hesitant to include this change in this PR -- would you be willing to restore the old behavior in this PR and then make a follow on PR that proposes fixing the behavior?

It would be nice to be able to keep the concerns (bug fix and buffer reduction) separate

File: parquet/src/arrow/arrow_writer/mod.rs [hunk 15]
```diff
@@ -152,43 +151,69 @@ impl<W: Write> ArrowWriter<W> {
...
 
     /// Flushes all buffered rows into a new row group
     pub fn flush(&mut self) -> Result<()> {
-        self.flush_rows(self.buffered_rows)
+        let in_progress = match self.in_progress.take() {
+            Some(in_progress) => in_progress,
+            None => return Ok(()),
+        };
+
+        self.buffered_rows = 0;
```
Comment (alamb) [inline parquet/src/arrow/arrow_writer/mod.rs:210]: Can `self.buffered_rows` get out of sync here (like can it be left at zero if, for example, `next_row_group()` returns and error?

**Review by alamb — COMMENTED** [review]
Thank you very much @tustvold  -- this looks really nice 👌 

I am not sure about the `num_values()` metadata change - - I recommend we make that a separate PR. 

Otherwise I think it looks awesome -- I will verify the memory usage using the reproducer in https://github.com/influxdata/influxdb_iox/pull/7855

File: parquet/src/arrow/arrow_writer/mod.rs (hunk 13 above)
Comment (tustvold) [reply parquet/src/arrow/arrow_writer/mod.rs:302]: I tried a lot to avoid this, however, I couldn't devise any mechanism that wasn't both a breaking change and pretty grim. This was the least bad option, and doesn't commit us to anything long term

File: parquet/src/arrow/arrow_writer/mod.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (tustvold) [reply parquet/src/arrow/arrow_writer/mod.rs:155]: > to be more consistent with 

RecordBatch (and Array) I think is the outlier in this regard:

* https://docs.rs/arrow-schema/latest/arrow_schema/struct.Fields.html#method.size
* https://docs.rs/arrow-row/latest/arrow_row/?search=size
* https://docs.rs/arrow/latest/arrow/datatypes/struct.UnionFields.html#method.size

I would rather keep the current name, especially as it technically isn't really the memory size but specifically the row group size

File: parquet/src/column/writer/mod.rs (hunk 8 above)
Comment (tustvold) [reply parquet/src/column/writer/mod.rs:1531]: I will split it out into a separate PR, but keep it here as otherwise I need to add logic to ArrowPageWriter to reproduce what I'm certain is an incorrect behaviour

Comment (tustvold) [reply parquet/src/column/writer/mod.rs:1531]: PR here - https://github.com/apache/arrow-rs/pull/4285

Turns out it is a testing quirk, there is no behaviour change

File: parquet/src/arrow/arrow_writer/mod.rs [hunk 16]
```diff
@@ -152,43 +147,75 @@ impl<W: Write> ArrowWriter<W> {
         self.writer.flushed_row_groups()
     }
 
-    /// Enqueues the provided `RecordBatch` to be written
+    /// Returns the length in bytes of the current in progress row group
```
Comment (alamb) [inline parquet/src/arrow/arrow_writer/mod.rs:150]:
```suggestion
    /// The bytes of memory used to store the current in progress row group.
    ///
    /// The writer incrementally writes each row group  on a call to `write` to
    /// a buffer in memory. This function reports on the amount of memory
    /// in the buffer.
```

**Review by alamb — COMMENTED** [review]
Looking good.



In order to get DataFusion to compile when testing this PR, I had to add some additional `Send` bounds: https://github.com/apache/arrow-rs/commit/00028c95d730c466ea52dc2c7d558bb7204afa83

You can see the specific error here: https://github.com/influxdata/influxdb_iox/pull/7866#discussion_r1205910780

Is it possible to add the `Send` bounds (or do a better fix) in this PR?

**Comment by tustvold** [comment]
https://github.com/apache/arrow-rs/pull/4287 adds the necessary Send bounds, although I am also exploring an approach that doesn't need them

File: parquet/src/arrow/arrow_writer/mod.rs [hunk 17]
```diff
@@ -152,43 +147,75 @@ impl<W: Write> ArrowWriter<W> {
...
+    pub fn in_progress_size(&self) -> usize {
```
[the remark this answers is not in the document]
Comment (tustvold) [reply parquet/src/arrow/arrow_writer/mod.rs:151]: Aah, this is an oversight on my part. The reported size will just be the size of the flushed pages, any data buffered but not yet flushed to a page will not be counted. This should be at most 1 page per column, and so should be less than the max page size (1MB) per column

[the remark this answers is not in the document]
Comment (alamb) [reply parquet/src/arrow/arrow_writer/mod.rs:151]: Can we make this API return the upper bound? Otherwise I am not sure how useful it is 🤔

[the remark this answers is not in the document]
Comment (alamb) [reply parquet/src/arrow/arrow_writer/mod.rs:151]: To be clear here, the usecase is "I want to ensure that when writing many parquet files concurrently we don't exceed some memory limit (so that the process doing so isn't killed by k8s / the operating system)

This doesn't need to be super accurate, but enough of an upper bound to achieve the above goal.

If it would be ok to add a 1MB overhead for each column (e.g. the PAGE_SIZE)or wherever that buffer is defined, I can try and propose a patch to do so.

[the remark this answers is not in the document]
Comment (tustvold) [reply parquet/src/arrow/arrow_writer/mod.rs:151]: I need to think about how best to support an upper bound, we don't currently expose this information. I had assumed, perhaps incorrectly, that an order of magnitude value was sufficient. Ultimately if your at a point where an individual page matters, you're going to have problems...

[the remark this answers is not in the document]
Comment (tustvold) [reply parquet/src/arrow/arrow_writer/mod.rs:151]: i've inclued https://github.com/tustvold/arrow-rs/pull/63 in https://github.com/apache/arrow-rs/pull/4280/commits/4683a301048d826344dc5d483891997928ffe446 along with a more accurate form of memory estimation - PTAL

File: parquet/src/column/writer/mod.rs [hunk 18]
```diff
@@ -441,6 +452,16 @@ impl<'a, E: ColumnValueEncoder> GenericColumnWriter<'a, E> {
         self.column_metrics.total_bytes_written
     }
 
+    /// Returns the estimated total bytes for this column writer
+    ///
+    /// Unlike [`Self::get_total_bytes_written`] this includes an estimate
+    /// of any data that has not yet been flushed to a pge
+    pub(crate) fn get_estimated_total_bytes(&self) -> u64 {
+        self.column_metrics.total_bytes_written
+            + self.encoder.estimated_data_page_size() as u64
```
Comment (tustvold) [inline parquet/src/column/writer/mod.rs:461]: As the encoders all buffer encoded data internally, this is a fairly accurate estimation of the memory usage.

File: parquet/src/column/writer/mod.rs [hunk 19]
```diff
@@ -421,10 +449,24 @@ impl<'a, E: ColumnValueEncoder> GenericColumnWriter<'a, E> {
...
     pub fn get_total_bytes_written(&self) -> u64 {
         self.column_metrics.total_bytes_written
     }
 
+    /// Returns the estimated total bytes for this column writer
+    ///
+    /// Unlike [`Self::get_total_bytes_written`] this includes an estimate
+    /// of any data that has not yet been flushed to a page
+    #[cfg(feature = "arrow")]
+    pub(crate) fn get_estimated_total_bytes(&self) -> u64 {
```
Comment (tustvold) [inline parquet/src/column/writer/mod.rs:464]: I've left this pub(crate) as I'm not sure about the returned value, it is potentially misleading as it doesn't include the impact of any block compression

**Review by alamb — APPROVED** [review]
Thank you @tustvold  -- I don't have time today to pull this all together and try it inside IOx, but I will do so tomorrow.

**Comment by alamb** [comment]
FYI @sundy-li
